src/IR.hs

```diff
@@ -1,43 +1,96 @@
-import           IR.LambdaLift                  ( liftProgramLambdas )
 
-lowerAst :: A.Program -> Compiler.Pass (I.Program Ann.Type)
-lowerAst = lowerProgram
+import           Control.Monad                  ( when )
+import           System.Console.GetOpt          ( ArgDescr(..)
+                                                , OptDescr(..)
+                                                )
+
+-- | Operation modes for the IR compiler stage.
+data Mode
+  = DumpIR
+  | DumpIRTyped
+  | DumpIRLifted
+  | DumpIRFinal
+  | Continue
+  deriving (Eq, Show)
+
+-- | Compiler options for the IR compiler stage.
+newtype Options = Options { mode :: Mode } deriving (Eq, Show)
+
+instance Default Options where
+  def = Options { mode = Continue }
 
-inferTypes :: I.Program Ann.Type -> Compiler.Pass (I.Program Classes.Type)
-inferTypes = inferProgram
+-- | CLI options for the IR compiler stage.
+options :: [OptDescr (Options -> Options)]
+options =
+  [ Option ""
+           ["dump-ir"]
+           (NoArg $ setMode DumpIR)
+           "Print the IR immediately after lowering"
+  , Option ""
+           ["dump-ir-typed"]
+           (NoArg $ setMode DumpIRTyped)
+           "Print the fully-typed IR after type inference"
+  , Option ""
+           ["dump-ir-final"]
+           (NoArg $ setMode DumpIRFinal)
+           "Print the last IR representation before code generation"
+  ]
+ where setMode m o = o { mode = m }
 
-instantiateClasses
-  :: I.Program Classes.Type -> Compiler.Pass (I.Program Poly.Type)
-instantiateClasses = instProgram
+-- | IR compiler sub-stage, lowering AST to optionally type-annotated IR.
+lower :: Options -> A.Program -> Pass (I.Program Ann.Type)
+lower opt prg = do
+  ir <- lowerProgram prg
+  when (mode opt == DumpIR) $ dump ir
+  return ir
 
-yieldAbstraction :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
-yieldAbstraction = return
+-- | IR compiler sub-stage, ultimately producing a fully-typed IR.
+ann2Class :: Options -> I.Program Ann.Type -> Pass (I.Program Cls.Type)
+ann2Class opt ir = do
+  irInferred <- inferProgram ir
+  when (mode opt == DumpIRTyped) $ dump irInferred
+  return irInferred
 
-lambdaLift :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
-lambdaLift = liftProgramLambdas
+-- | IR compiler sub-stage, ultimately instantiating all type classes.
+class2Poly :: Options -> I.Program Cls.Type -> Pass (I.Program Poly.Type)
+class2Poly _ = instProgram
 
-defunctionalize :: I.Program Poly.Type -> Compiler.Pass (I.Program Poly.Type)
-defunctionalize = return
+{- | IR compiler sub-stage, ultimately monomorphizing the polymorphic type
+system.
 
-inferDrops :: I.Program a -> Compiler.Pass (I.Program a)
-inferDrops = return
+TODO: This is probably not necessary in the future.
+-}
+poly2Flat :: Options -> I.Program Poly.Type -> Pass (I.Program Flat.Type)
+poly2Flat opt ir = do
+  irLifted <- liftProgramLambdas ir
+  when (mode opt == DumpIRLifted) $ dump irLifted
```

Comment:
Why is lifting not its own operation here?

Reply:
Like why is it no longer a top-level operation?

Reply:
Why does it need to be? I'm encapsulating it here so that there doesn't need to be all this extra boilerplate invoking all the subphases.

Reply:
I guess it doesn't have to be, I suppose I just saw it that why because I implemented it. I see now why your organization makes sense, just wanted to ask why this change was made.

Reply:
Gotcha. Yeah I did that because I saw that there was a lot of boilerplate all over the place, stitching together the same compiler pipeline in different places.